Fix bug involving null columns.

deusnefum opened the pull request:
In using the vertica driver, I was getting a "syntax invalid" error trying to rows.Scan() a column that had a null value.

Looking at rows.go, I think the use of sql.NullInt64, et al types were not used correctly. The sql.NullX types are structs with the type plus a bool to indicate if it's null or not. The way they're used in rows.go, it looks like it's used as though it *means* the field is null. 

The database/sql interface expects you to either set your columns to nil or set them to their actual value and it handles conversion to an sql.NullX type.

I made a little test program to demonstrate:


```go
package main

import (
	"database/sql"
	"fmt"
	//	_ "github.com/vertica/vertica-sql-go"
	_ "github.com/deusnefum/vertica-sql-go"
)

func main() {
	db, err := sql.Open("vertica", "vertica://dbadmin:SuperFantasticPassword@localhost:5433/dbadmin")
	if err != nil {
		panic(err)
	}

	rows, err := db.Query("SELECT 'tom' AS name, null AS id, null AS ncol")
	if err != nil {
		panic(err)
	}

	var n string
	var i sql.NullInt64
	var q sql.NullString

	for rows.Next() {
		rows.Scan(&n, &i, &q)
	}

	fmt.Println(n, i, q)
}
```

When I use the upstream code I get this:

> tom {0 true} { false}

The second col, id, was null. The value should've been {0 false}.

When I used my fixed code:

>  tom {0 false} { false}

I also ran the driver_test.go file and all passed except for an error with testSTDINCopy, which I don't think is related to my change. To get the flags to work, I had to add flags.Parse() to init, which fixes https://github.com/vertica/vertica-sql-go/issues/33.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vertica/vertica-sql-go?pullRequest=37) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vertica/vertica-sql-go?pullRequest=37) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vertica/vertica-sql-go?pullRequest=37) it.</sub>

deusnefum commented:
I see the Travis CI build failed. Can you get me logs? This builds fine on my system with go 1.13.

sitingren commented:
> I see the Travis CI build failed. Can you get me logs? This builds fine on my system with go 1.13.

https://travis-ci.com/vertica/vertica-sql-go/jobs/239980101#L641

deusnefum commented:
I did a bunch of tests with go 1.13, 1.12.11, and 1.11.

I don't get the error (FAIL: TestValueTypes) testing against our vertica test cluster, which is running 7.x. I used the travis CI logs to figure out how to replicate the docker container with 9.x and was able to recreate the fail.

Looking closer at driver_test.go, it sensibly made the same assumptions about how the sql.NullX types work. I changed the test to match the code.

I got another fail from the CopyFromSTDIN / FromStream tests, so I modified the query to force a sort-by-name and it started passing.

Switching between 1.11, 1.12, and 1.13 really underlined how annoying that breaking change is. I've commented out flags.Parse() and testing.Init(), but noted they may be necessary.

huebnerr commented:
Thanks for the updates. I'm traveling right now, but will catch up shortly. In the mean time, Siting Ren, can cover for me.

Great work. Yeah, a real pain about those breaking changes in the testing harness. In 1.11 and 1.12, it was much easier. :)

On Sep 30, 2019 2:53 PM, Matthew Hiles <notifications@github.com> wrote:

I did a bunch of tests with go 1.13, 1.12.11, and 1.11.

I don't get the error (FAIL: TestValueTypes) testing against our vertica test cluster, which is running 7.x. I used the travis CI logs to figure out how to replicate the docker container with 9.x and was able to recreate the fail.

Looking closer at driver_test.go, it sensibly made the same assumptions about how the sql.NullX types work. I changed the test to match the code.

I got another fail from the CopyFromSTDIN / FromStream tests, so I modified the query to force a sort-by-name and it started passing.

Switching between 1.11, 1.12, and 1.13 really underlined how annoying that breaking change is. I've commented out flags.Parse() and testing.Init(), but noted they may be necessary.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/vertica/vertica-sql-go/pull/37?email_source=notifications&email_token=AK56SCJIKEDW5XOLMP3STZTQMJKMBA5CNFSM4I3LOJB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD764IMQ#issuecomment-536724530>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AK56SCJNWJ3YKI56W3Z27SLQMJKMBANCNFSM4I3LOJBQ>.

huebnerr commented:
Reviewed and gratefully accepted.
